Add Check Support for Java 21 Pattern Matching for Switch Syntax: SimplifyBooleanExpression

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation:https://checkstyle.org/checks/coding/simplifybooleanexpression.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="SimplifyBooleanExpression">

        </module>
    </module>
</module>
➜  src cat Test.java  
record ColoredPoint(boolean p, int x, String c) { }
record Rectangle(boolean upperLeft, boolean lowerRight) { }
public class Test {

    void test(Object obj) {
        switch (obj) {
            case ColoredPoint(boolean x,_,_) when (x == true):   // violation
                System.out.println("coloredPoint");
                break;
            case Rectangle(boolean y,_) when (!(y != true)) : // violation
                System.out.println("Rectangle");
                break;
            case Rectangle(boolean y,boolean z) when (z == false) : // violation
                System.out.println("Rectangle");
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + obj);
        }
    }
}
➜  src javac --enable-preview --release 21 Test.java
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:7:54: Expression can be simplified. [SimplifyBooleanExpression]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:10:51: Expression can be simplified. [SimplifyBooleanExpression]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:13:57: Expression can be simplified. [SimplifyBooleanExpression]
Audit done.
Checkstyle ends with 3 errors.

```

Describe what you want in detail

we are ok here. we can close or approve for input file.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanExpressionCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanExpressionCheckTest.java
@@ -56,6 +56,19 @@ public class SimplifyBooleanExpressionCheckTest
     }
 
     @Test
+    public void testWhenExpression() throws Exception {
+        final String[] expected = {
+            "13:41: " + getCheckMessage(MSG_KEY),
+            "16:41: " + getCheckMessage(MSG_KEY),
+            "17:41: " + getCheckMessage(MSG_KEY),
+            "22:41: " + getCheckMessage(MSG_KEY),
+            "23:44: " + getCheckMessage(MSG_KEY),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputSimplifyBooleanExpressionWithWhen.java"), expected);
+    }
+
+    @Test
     public void testTokensNotNull() {
         final SimplifyBooleanExpressionCheck check = new SimplifyBooleanExpressionCheck();
         assertWithMessage("Acceptable tokens should not be null")

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/simplifybooleanexpression/InputSimplifyBooleanExpressionWithWhen.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/simplifybooleanexpression/InputSimplifyBooleanExpressionWithWhen.java
@@ -1,0 +1,46 @@
+/*
+SimplifyBooleanExpression
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.simplifybooleanexpression;
+
+public class InputSimplifyBooleanExpressionWithWhen {
+
+    void test(Object o) {
+        if (o instanceof Boolean b && b == true) { } // violation
+
+        switch (o) {
+            case R(boolean x, _) when x == true -> {} // violation
+            case R(_, boolean y) when y != false -> {} // violation
+            default -> {}
+        }
+
+        switch (o) {
+            case R(boolean x, _) when x == false -> {} // violation
+            case R(_, boolean y) when (!(y != true)) -> {} // violation
+            default -> {}
+        }
+    }
+
+    void test2(Object o) {
+
+        if (o instanceof Boolean b && b) { }
+
+        switch (o) {
+            case R(boolean x, _) when x -> {}
+            case R(_, boolean y) when y -> {}
+            default -> {}
+        }
+
+        switch (o) {
+            case R(boolean x, _) when !x -> {}
+            case R(_, boolean y) when y -> {}
+            default -> {}
+        }
+
+    }
+    record R(boolean x, boolean y) {}
+}

